### PR TITLE
Add half precision testing [1/n]

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Issue https://github.com/PyTorchLightning/metrics/issues/57
Adds testing interface for half precision and apply to few metrics.
Follow up PRs will add the testing to remaining metrics and add documentation to metrics where it will not work.

Note, that we cannot really do anything about it not working for some of the metrics as some pytorch operations (e.g `torch.log`) does not support half precision + cpu.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
can we rather use this function version? https://github.com/PyTorchLightning/pytorch-lightning/pull/6434

**Comment by SkafteNicki** [comment]
> 
> 
> can we rather use this function version? [PyTorchLightning/pytorch-lightning#6434](https://github.com/PyTorchLightning/pytorch-lightning/pull/6434)

Of course :]

File: tests/helpers/testers.py [hunk 1]
```diff
@@ -167,6 +167,41 @@ def _functional_test(
...
+    Test if an metric can be used with half precision tensors
+
+    Args:
+        metric_module: the metric module to test
+        metric_functional: the metric functional to test
+        preds: torch tensor with predictions
+        target: torch tensor with targets
+        device: determine device, either "cpu" or "cuda"
+    """
+    if device == 'cpu':
```
Comment (tchaton) [inline tests/helpers/testers.py:187]: Can you merge them as only `.cuda()` is being added.

File: torchmetrics/functional/regression/mean_squared_error.py [hunk 2]
```diff
@@ -20,7 +20,8 @@
 
 def _mean_squared_error_update(preds: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, int]:
     _check_same_shape(preds, target)
-    sum_squared_error = torch.sum(torch.pow(preds - target, 2))
+    diff = preds - target
+    sum_squared_error = torch.sum(diff * diff)
```
Comment (tchaton) [inline torchmetrics/functional/regression/mean_squared_error.py:24]: Do you know if there is a speed impact between diff * diff and torch.pow ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: torchmetrics/functional/regression/mean_squared_error.py (hunk 2 above)
Comment (SkafteNicki) [reply torchmetrics/functional/regression/mean_squared_error.py:24]: diff * diff is approximately twice as fast compared to torch.pow (with a little variation depending on the size of the input).

File: tests/helpers/testers.py (hunk 1 above)
Comment (SkafteNicki) [reply tests/helpers/testers.py]: done :]

File: torchmetrics/functional/regression/mean_squared_error.py (hunk 2 above)
Comment (justusschock) [reply torchmetrics/functional/regression/mean_squared_error.py:24]: That's only true for torch.pow(..., 2). For higher powers, the pow function is considererd to be faster

File: tests/regression/test_mean_error.py [hunk 3]
```diff
@@ -92,6 +93,20 @@ def test_mean_error_functional(self, preds, target, sk_metric, metric_class, met
             sk_metric=partial(sk_metric, sk_fn=sk_fn),
         )
 
+    @pytest.mark.skipif(
+        not _TORCH_GREATER_EQUAL_1_6,
```
Comment (justusschock) [inline tests/regression/test_mean_error.py:97]: Not sure if we really should skip here or show a more meaningful error message.

**Review by justusschock — APPROVED** [review]

File: tests/regression/test_mean_error.py (hunk 3 above)
Comment (Borda) [reply tests/regression/test_mean_error.py:97]: @SkafteNicki ^^

**Comment by Borda** [comment]
@SkafteNicki mind resolve conflicts... :]

File: torchmetrics/utilities/__init__.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
 from torchmetrics.utilities.data import apply_to_collection  # noqa: F401
 from torchmetrics.utilities.distributed import class_reduce, reduce  # noqa: F401
+from torchmetrics.utilities.imports import _TORCH_GREATER_EQUAL_1_6  # noqa: F401
```
Comment (Borda) [inline torchmetrics/utilities/__init__.py:3]:
```suggestion
```

File: tests/regression/test_mean_error.py [hunk 5]
```diff
@@ -23,6 +23,7 @@
 from tests.helpers.testers import BATCH_SIZE, NUM_BATCHES, MetricTester
 from torchmetrics.functional import mean_absolute_error, mean_squared_error, mean_squared_log_error
 from torchmetrics.regression import MeanAbsoluteError, MeanSquaredError, MeanSquaredLogError
+from torchmetrics.utilities import _TORCH_GREATER_EQUAL_1_6
```
Comment (Borda) [inline tests/regression/test_mean_error.py:26]:
```suggestion
from torchmetrics.utilities.imports import _TORCH_GREATER_EQUAL_1_6
```

File: tests/regression/test_mean_error.py (hunk 3 above)
Comment (SkafteNicki) [reply tests/regression/test_mean_error.py:99]: @justusschock you got any proposal?

Comment (ananyahjha93) [reply tests/regression/test_mean_error.py:99]: @justusschock more meaningful message in test or the class itself? wouldn't be easier to test for half precision tensors in the base metric class and check if torch version >= 1.6

**Review by ananyahjha93 — APPROVED** [review]

File: tests/helpers/testers.py [hunk 6]
```diff
@@ -170,6 +170,30 @@ def _functional_test(
...
+        metric_functional: the metric functional to test
+        preds: torch tensor with predictions
+        target: torch tensor with targets
+        device: determine device, either "cpu" or "cuda"
+    """
+    p = preds[0].half().to(device) if preds[0].is_floating_point() else preds[0].to(device)
+    t = target[0].half().to(device) if target[0].is_floating_point() else target[0].to(device)
+    metric_module = metric_module.to(device)
+    assert metric_module(p, t)
+    assert metric_functional(p, t)
```
Comment (Borda) [inline tests/helpers/testers.py:194]:
```suggestion
    y_hat = preds[0].half().to(device) if preds[0].is_floating_point() else preds[0].to(device)
    y = target[0].half().to(device) if target[0].is_floating_point() else target[0].to(device)
    metric_module = metric_module.to(device)
    assert metric_module(y_hat, y)
    assert metric_functional(y_hat, y)
```

**Review by Borda — APPROVED** [review]
